src/Microsoft.AspNetCore.Mvc.TagHelpers/Cache/CacheTagKey.cs

```diff
@@ -0,0 +1,358 @@
+
+using System;
+using System.Collections.Generic;
+using System.Security.Cryptography;
+using System.Text;
+using Microsoft.AspNetCore.Razor.TagHelpers;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.AspNetCore.Mvc.TagHelpers.Cache
+{
+    /// <summary>
+    /// An instance of <see cref="CacheTagKey"/> represents the state of <see cref="CacheTagHelper"/>
+    /// or <see cref="DistributedCacheTagHelper"/> keys.
+    /// </summary>
+    public class CacheTagKey : IEquatable<CacheTagKey>
+    {
+        private static readonly char[] AttributeSeparator = new[] { ',' };
+
+        private const string CacheKeyTokenSeparator = "||";
+        private const string VaryByName = "VaryBy";
+        private const string VaryByHeaderName = "VaryByHeader";
+        private const string VaryByQueryName = "VaryByQuery";
+        private const string VaryByRouteName = "VaryByRoute";
+        private const string VaryByCookieName = "VaryByCookie";
+        private const string VaryByUserName = "VaryByUser";
+
+        private string _key;
+        private string _prefix;
+        private string _varyBy;
+        private DateTimeOffset? _expiresOn;
+        private TimeSpan? _expiresAfter;
+        private TimeSpan? _expiresSliding;
+        private IList<KeyValuePair<string, string>> _headers;
+        private IList<KeyValuePair<string, string>> _queries;
+        private IList<KeyValuePair<string, string>> _routeValues;
+        private IList<KeyValuePair<string, string>> _cookies;
+        private bool _varyByUser;
+        private string _username;
+
+        private string _generatedKey;
+        private int? _hashcode;
+        
+        private CacheTagKey()
+        {
+        }
+
+        /// <summary>
+        /// Creates an instance of <see cref="CacheTagKey"/> for a specific <see cref="CacheTagHelper"/>.
+        /// </summary>
+        /// <param name="tagHelper">The <see cref="CacheTagHelper"/>.</param>
+        /// <param name="context">The <see cref="TagHelperContext"/>.</param>
+        /// <returns>A new <see cref="CacheTagKey"/>.</returns>
+        public static CacheTagKey From(CacheTagHelper tagHelper, TagHelperContext context)
+        {
+            var cacheKey = FromCore(tagHelper, context);
+
+            cacheKey._key = context.UniqueId;
+            cacheKey._prefix = nameof(CacheTagHelper);
+
+            return cacheKey;
+        }
+
+        /// <summary>
+        /// Creates an instance of <see cref="CacheTagKey"/> for a specific <see cref="DistributedCacheTagHelper"/>.
+        /// </summary>
+        /// <param name="tagHelper">The <see cref="DistributedCacheTagHelper"/>.</param>
+        /// <param name="context">The <see cref="TagHelperContext"/>.</param>
+        /// <returns>A new <see cref="CacheTagKey"/>.</returns>
+        public static CacheTagKey From(DistributedCacheTagHelper tagHelper, TagHelperContext context)
```

Comment:
Seems like an odd API name. Couldn't we have regular ctors for this instead of factories?